### PR TITLE
improvement: reduce number of state clones

### PR DESCRIPTION
**Pull request by Wodann**
This contains both EDR and Hardhat changes, so requires review from both sides.

### CODE REVIEW

File: packages/hardhat-core/src/internal/hardhat-network/provider/node.ts [hunk 1]
```diff
@@ -1924,25 +1908,23 @@ export class HardhatNode extends EventEmitter {
...
 
-    const block = await this.getBlockByNumber(blockNumberOrPending);
-    if (block === undefined) {
-      // TODO handle this better
-      throw new Error(
-        `Block with number ${blockNumberOrPending.toString()} doesn't exist. This should never happen.`
-      );
-    }
+    // const currentStateRoot = await this._context.vm().getStateRoot();
+    // const snapshotId = await this._context.vm().makeSnapshot();
```
Comment (agostbiro) [inline packages/hardhat-core/src/internal/hardhat-network/provider/node.ts:1917]: Can the two lines above be removed?

File: packages/hardhat-core/src/internal/hardhat-network/provider/node.ts [hunk 2]
```diff
@@ -1924,25 +1908,23 @@ export class HardhatNode extends EventEmitter {
...
 
-    const snapshot = await this._context.vm().makeSnapshot();
-    await this._setBlockContext(block);
     try {
       return await action();
     } finally {
-      await this._context.vm().restoreContext(snapshot);
-      await this._context.vm().removeSnapshot(snapshot);
+      await this._context.vm().restoreBlockContext(latestBlockNumber);
+      // await this._context.vm().restoreSnapshot(snapshotId);
```
Comment (agostbiro) [inline packages/hardhat-core/src/internal/hardhat-network/provider/node.ts:1924]: Can this line be removed?

File: crates/rethnet_evm/src/state/fork.rs [hunk 3]
```diff
@@ -79,6 +76,14 @@ impl ForkState {
         self.remote_state.lock().set_block_number(block_number);
     }
 
+    /// Sets the current state root, preventing it from being regenerated.
+    pub fn set_state_root(&mut self, state_root: &B256) {
+        let local_state_root = self.local_state.state_root().unwrap();
```
Comment (agostbiro) [inline crates/rethnet_evm/src/state/fork.rs:81]: Is this guaranteed to exist? Would prefer an `expect` if yes

File: packages/hardhat-core/src/internal/hardhat-network/provider/node.ts (hunk 1 above)
Comment (Wodann) [reply packages/hardhat-core/src/internal/hardhat-network/provider/node.ts]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4418/commits/9281dc1cb7410eb2ec79d38c66864033fd83542d

File: packages/hardhat-core/src/internal/hardhat-network/provider/node.ts (hunk 2 above)
Comment (Wodann) [reply packages/hardhat-core/src/internal/hardhat-network/provider/node.ts]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4418/commits/9281dc1cb7410eb2ec79d38c66864033fd83542d

File: crates/rethnet_evm/src/state/fork.rs (hunk 3 above)
Comment (Wodann) [reply crates/rethnet_evm/src/state/fork.rs:81]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4418/commits/162b014949e5f86b06eb6bd6111028ee9d934de5

**Comment by Wodann** [comment]
Remove the leftover line: https://github.com/NomicFoundation/hardhat/pull/4418/commits/162b014949e5f86b06eb6bd6111028ee9d934de5

**Comment by agostbiro** [comment]
Thanks for the changes @Wodann . It looks like there are some more CI failures that are concerning. In particular:

- [lint](https://github.com/NomicFoundation/hardhat/actions/runs/6318212352/job/17156723814?pr=4418)
- [different state root in dual mode](https://github.com/NomicFoundation/hardhat/actions/runs/6318212351/job/17156765411?pr=4418)
- [different state in hardhat-ethers](https://github.com/NomicFoundation/hardhat/actions/runs/6318212349/job/17156745013?pr=4418)

**Comment by agostbiro** [comment]
Thanks for this difficult PR @Wodann ! There are some merge conflicts that need to be solved and could you write a few words please about the problem this PR solves and the approach taken to help with the review?

File: crates/edr_evm/src/blockchain.rs [hunk 4]
```diff
@@ -130,10 +130,16 @@ pub trait Blockchain {
...
+    /// Retrieves the state at a given block.
+    ///
+    /// The state overrides are applied after the block they are associated
+    /// with. The specified override of a nonce may be ignored to maintain
+    /// validity.
     async fn state_at_block_number(
         &self,
         block_number: u64,
+        // Block number -> state overrides
+        state_overrides: &BTreeMap<u64, StateOverride>,
```
Comment (Wodann) [inline crates/edr_evm/src/blockchain.rs:142]: When constructing a state at a block number, it's possible that someone wants to override the state at a particular block. An example of this are Hardhat's irregular state changes. These need to be passed in to ensure that they're applied in the same order.

E.g.

Genesis block: no transactions, so no state committed
Irregular: genesis accounts are added
Block 1: for all transactions combined, commit state
Block 2: for all transactions combined, commit state
Irregular: set_code & set_balance are called
Block 3: for all transactions combined, commit state

To retrieve the state, these need to be applied in the same order, including the state overrides (irregular state changes).

Previously we weren't doing this, resulting in incorrect state construction. This resolves state root inequality between HH & EDR, which we were seeing in the past.

File: packages/hardhat-core/src/internal/hardhat-network/provider/EdrState.ts [hunk 5]
```diff
@@ -29,36 +27,6 @@ export class EdrStateManager {
     );
   }
 
-  public static async forkRemote(
```
Comment (Wodann) [inline packages/hardhat-core/src/internal/hardhat-network/provider/EdrState.ts:32]: Remove unused function

File: packages/hardhat-core/src/internal/hardhat-network/provider/blockchain/edr.ts [hunk 6]
```diff
@@ -13,10 +13,12 @@ import {
 import { FilterParams } from "../node-types";
 import { bloomFilter, filterLogs } from "../filter";
 import { Bloom } from "../utils/bloom";
+import { EdrIrregularState } from "../EdrIrregularState";
 
 export class EdrBlockchain implements BlockchainAdapter {
   constructor(
     private readonly _blockchain: Blockchain,
+    private readonly _irregularState: EdrIrregularState,
```
Comment (Wodann) [inline packages/hardhat-core/src/internal/hardhat-network/provider/blockchain/edr.ts:21]: Maintain the irregular state for usage in retrieving a state at a particular block

File: packages/hardhat-core/src/internal/hardhat-network/provider/node.ts [hunk 7]
```diff
@@ -1182,16 +1169,14 @@ export class HardhatNode extends EventEmitter {
   ): Promise<void> {
     const account = await this._context.vm().getAccount(address);
     account.balance = newBalance;
-    await this._context.vm().putAccount(address, account);
-    await this._persistIrregularWorldState();
```
Comment (Wodann) [inline packages/hardhat-core/src/internal/hardhat-network/provider/node.ts:1186]: Persisting irregular world state is handled internally to the VM adapter

File: packages/hardhat-core/src/internal/hardhat-network/provider/vm/block-builder/hardhat.ts [hunk 8]
```diff
@@ -52,6 +51,11 @@ export class HardhatBlockBuilder implements BlockBuilderAdapter {
       );
     }
 
+    if (!this._checkpointed) {
```
Comment (Wodann) [inline packages/hardhat-core/src/internal/hardhat-network/provider/vm/block-builder/hardhat.ts:54]: This restores the code to what was originally happening on `main`

File: packages/hardhat-core/src/internal/hardhat-network/provider/vm/edr.ts [hunk 9]
```diff
@@ -88,46 +93,6 @@ export class EdrAdapter implements VMAdapter {
     this._vmTracer = new VMTracer(_common, false);
   }
 
-  public static async create(
```
Comment (Wodann) [inline packages/hardhat-core/src/internal/hardhat-network/provider/vm/edr.ts:91]: No longer needed, as we always create the VM in the `Context` class

File: packages/hardhat-core/src/internal/hardhat-network/provider/vm/rethnet.ts [hunk 10]
```diff
@@ -205,18 +190,18 @@ export class RethnetAdapter implements VMAdapter {
       }
     );
 
-    this._stateRootToState.set(
-      await this.getStateRoot(),
-      await this._state.asInner().deepClone()
-    );
+    if (isIrregularChange === true) {
+      await this._persistIrregularAccount(address, modifiedAccount);
```
[the remark this answers is not in the document]
Comment (Wodann) [reply packages/hardhat-core/src/internal/hardhat-network/provider/vm/rethnet.ts]: The `RethnetStateManager` isn't publicly exposed; it's only used inside the VM. The caller of this function determines whether the state change is irregular or not, hence why this needs to be exposed here.

File: packages/hardhat-core/src/internal/hardhat-network/provider/vm/vm-adapter.ts [hunk 11]
```diff
@@ -48,22 +48,52 @@ export interface VMAdapter {
...
-  // setters
-  putAccount(address: Address, account: Account): Promise<void>;
-  putContractCode(address: Address, value: Buffer): Promise<void>;
+  /**
+   * Update the account info for the given address.
+   */
+  putAccount(
+    address: Address,
+    account: Account,
+    isIrregularChange?: boolean
```
[the remark this answers is not in the document]
Comment (Wodann) [reply packages/hardhat-core/src/internal/hardhat-network/provider/vm/vm-adapter.ts]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4418/commits/cf0805c62c2bea54b1822f4c88682d0aa326665a

**Pull request merged**